### PR TITLE
Necromancing CS1.6 support from cs branch

### PR DESCRIPTION
**Pull request by khanghugo**
Expansion of https://github.com/YaLTeR/BunnymodXT/pull/322

### CODE REVIEW

**Comment by khanghugo** [comment]
It seems really close. I am still figuring out why the friction does not behave as wished even though the line is there. The friction is not the only problem but also just moving on ground in general is still not accurately predicted so I guess it it could be something else.

File: src/hlstrafe.cpp [hunk 1]
```diff
@@ -1351,9 +1351,13 @@ namespace HLStrafe
 				player.Velocity[1] = static_cast<float>(std::sin(player.Viewangles[1] * M_DEG2RAD) * std::cos(player.Viewangles[0] * M_DEG2RAD) * 350 * 1.6);
 			}
 		}
-		player.Velocity[2] = static_cast<float>(std::sqrt(2 * 800 * 45.0));
+		player.Velocity[2] = static_cast<float>(std::sqrt(2 * 800 * 45.0) * ((100.0 - (player.StaminaTime / 1000.0) * 19.0) / 100.0));
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: This applies StaminaTime (which is the value of fuser2) unconditionally. However, other mods are free to put whatever unrelated value into fuser2, so there should be some sort of check to ignore it. Also below you set StaminaTime, also unconditionally, when it only applies to CS 1.6.

I'm not sure what the best way to do this, maybe a flag like HasStamina? Similar to the Bhopcap flag.

**Review by YaLTeR — COMMENTED** [review]
Another change is that in CS 1.6 +use doesn't incur a 0.333 slowdown I believe.

File: src/hlstrafe.cpp (hunk 1 above)
Comment (khanghugo) [reply src/hlstrafe.cpp]: Yeah, I can have the check similarly done in HwDLL and pass it into here through the MovementVars struct. I was really concerned with clogging the whole thing with conditional statements for just a single game but it seems like a okay.

**Comment by khanghugo** [comment]
Now it looks like everything works accordingly. Let me play around with it a bit and see if it is good for another review.

**Comment by khanghugo** [comment]
Friction still seems to weird out when the frame bulk is a little bit big enough. Just normal friction is okay but when lgagst is involved, it gets messed up by one frame, which only happens in long enough frame bulk. There is this instance where the it thinks it can moves on ground without being affected by friction. Interesting enough, the predicted velocity for that wrong is exactly how air velocity should be. So maybe there is some weird logic here and there. 

Let me just look for the differences between game physics again and see if that is fixable. Otherwise, the current form works pretty okay especially in addition of simulated client to rectify.

File: src/hlstrafe.cpp [hunk 2]
```diff
@@ -2031,10 +2043,13 @@ namespace HLStrafe
 		if (vars.Frametime != 0.f)
 			curState.PredictThis = State0ms::NOTHING;
 
-		bool reduceWishspeed = playerCopy.Ducking;
+		bool reduceWishspeed = playerCopy.Ducking || (vars.HasStamina && playerCopy.InDuckAnimation);
```
Comment (khanghugo) [inline src/hlstrafe.cpp]: What do you think the implication of reducing wishspeed when InDuckAnimation? I am having hard time wrapping around it as I have been comparing the physics but don't know for sure why this is included (in the original cs branch). Maybe I am missing something but removal of this does desync one made with it. In prediction, this slows down 1-2 unit than without, so there is some funky friction stuff.

**Comment by khanghugo** [comment]
> There is this instance where the it thinks it can moves on ground without being affected by friction

I have found out that it just happens when there is simulated client running. The normal prediction is fine. According to the prediction, they are a bit faster than what really happens. So there is still something not right about what is going on now.

File: src/hlstrafe.cpp [hunk 3]
```diff
@@ -1334,11 +1333,11 @@ namespace HLStrafe
 		}
 
 		if (vars.Bhopcap) {
-			auto maxscaledspeed = 1.7f * vars.Maxspeed;
+			auto maxscaledspeed = vars.SpeedScale * vars.Maxspeed;
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: Let's call this BhopcapMaxspeedScale

File: src/hlstrafe.cpp [hunk 4]
```diff
@@ -1334,11 +1333,11 @@ namespace HLStrafe
...
 			if (maxscaledspeed > 0) {
 				auto speed = Length<float, 3>(player.Velocity);
 				if (speed > maxscaledspeed)
-					VecScale<float, 3>(player.Velocity, (maxscaledspeed / speed) * 0.65, player.Velocity);
+					VecScale<float, 3>(player.Velocity, (maxscaledspeed / speed) * vars.BhopcapScale, player.Velocity);
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: And this BhopcapMultiplier

File: src/hlstrafe.cpp [hunk 5]
```diff
@@ -1256,7 +1256,6 @@ namespace HLStrafe
 			if (!curState.Duck && !player.Ducking) {
 				player.DuckTime = 1000;
 				player.InDuckAnimation = true;
-
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: Unrelated line removal

File: src/hlstrafe.cpp [hunk 6]
```diff
@@ -2032,27 +2042,40 @@ namespace HLStrafe
 			curState.PredictThis = State0ms::NOTHING;
 
 		bool reduceWishspeed = playerCopy.Ducking;
+
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: Unrelated line addition

File: src/hlstrafe.cpp [hunk 7]
```diff
@@ -2032,27 +2042,40 @@ namespace HLStrafe
...
 		// if this messes with Lgagst-ducktap() then it's the user's problem.
 		// (Though it shouldn't, generally).
-		if (out.Use && postype == PositionType::GROUND)
+		if (vars.UseSlow && out.Use && postype == PositionType::GROUND)
 			VecScale<float, 3>(playerCopy.Velocity, 0.3, playerCopy.Velocity);
 
 		LgagstJump(playerCopy, vars, postype, frame, out, reduceWishspeed, strafeButtons, useGivenButtons, curState, traceFunc, version);
 		Autojump(postype, frame, curState, out);
 		postype = PredictJump(playerCopy, postype, vars, frame, curState, out, traceFunc, true);
+
```
Comment (YaLTeR) [inline src/hlstrafe.cpp]: Unrelated line addition

File: src/hlstrafe.hpp [hunk 8]
```diff
@@ -37,6 +38,13 @@ namespace HLStrafe
 		float Stepsize;
 		float Bounce;
 		bool Bhopcap;
+		float BhopcapScale;
+		float SpeedScale;
+		bool UseSlow;
+
+		// Non-HL stuffs
```
Comment (YaLTeR) [inline src/hlstrafe.hpp]:
```suggestion
```

**Review by YaLTeR — CHANGES_REQUESTED** [review]
Looks good, just a few comments

**Comment by YaLTeR** [comment]
Thanks! Now in the BXT PR make the submodule point to the master HLStrafe commit.
